### PR TITLE
SI-6624 better lookup of case field accessors for case class pattern with complicated type

### PR DESCRIPTION
**Pull request by adriaanm**
(I need to run, but wanted to let the bot build this already -- will assign reviewer etc ASAP)

### CODE REVIEW

**Comment by adriaanm** [comment]
review by @retronym

**Comment by retronym** [comment]
lgtm

File: src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala [hunk 1]
```diff
@@ -417,7 +417,13 @@ trait PatternMatching extends Transform with TypingTransformers with ast.TreeDSL
...
             val treeMaker = TypeTestTreeMaker(patBinder, patBinder, extractor.paramType, extractor.paramType)(pos, extractorArgTypeTest = true)
             (List(treeMaker), treeMaker.nextBinder)
-          } else (Nil, patBinder)
+          } else {
+            // no type test needed, but the tree maker relies on `patBinderOrCasted` having type `extractor.paramType` (and not just some type compatible with it)
+            // SI-6624 shows this is necessary because apparently patBinder may have an unfortunate type (.decls don't have the case field accessors)
+            // TODO: get to the bottom of this -- I assume it happens when type checking infers a weird type for an unapply call
+            // by going back to the parameterType for the extractor call we get a saner type, so let's just do that for now
+            (Nil, patBinder setInfo extractor.paramType)
+          }
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/typechecker/PatternMatching.scala]: This seems like the kind of hinges-on-a-subtlety place where we would benefit from some logging. If you want to get to the bottom of this faster,

```
if (settings.developer.value) { // depends on another pull request
  if (!(patBinder.info =:= extractor.paramType))
    devWarning("say something useful...")
}
```

**Comment by adriaanm** [comment]
(the commit was only updated to improve docs/commit message)

**Pull request merged**